src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,589 @@
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#include <app/SimpleSubscriptionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace app {
+
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricIndexTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPeerNodeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kSubscriptionIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMinIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMaxIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricFilteredTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPathCountTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPathTypeTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEndpointIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kClusterIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kAttributeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEventIdTag;
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::SaveIndex(const SubscriptionIndex & index)
+{
+    if (index.mSize == 0)
+    {
+        return mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName());
+    }
+
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    backingBuffer.Calloc(MaxIndexSize());
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), MaxIndexSize());
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
+
+    for (size_t i = 0; i < index.mSize; ++i)
+    {
+        TLV::TLVType innerType;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, innerType));
+        ReturnErrorOnFailure(writer.Put(kFabricIndexTag, index[i].GetFabricIndex()));
+        ReturnErrorOnFailure(writer.Put(kPeerNodeIdTag, index[i].GetNodeId()));
+        ReturnErrorOnFailure(writer.EndContainer(innerType));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(arrayType));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName(),
+                                                   backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+#define DeleteIndexAndReturnLogErrorOnFailure(expr)                                                                                \
```

Comment:
This two-table implementation is a dangerous design.  Every place this macro is called is an opportunity to leak the scoped-node-id keyed subscription entries.

I would suggest that instead of the current design, this be implemented as a single kvstore table with a simple index from 0 to [1 - number of persisted subscriptions we wish to support].  Each entry can be fully self-describing.  When the node wakes, it attempt to read each entry and load those into memory that exist.

This would eliminate the risk of storage leaks, which can be catastrophic.

Reply:
I've changed the implementation as suggested.